### PR TITLE
pruntime: Support for deploy sidevm to selected workers.

### PR DESCRIPTION
**Pull request by kvinwang**
This PR add the capacity of deploying sidevm to selected workers to pRuntime/pink-runtime.

This enables the driver `sidevm_deployer` to support to deploy `paid` SideVM instances for any contract.

Contracts can deploy paid sidevm by cross calling the `SidevmOperation::deploy_to_workers` method with amount of transfered tokens.
The deployment can fail when the following condition met:
- Invalid arguments
- paid value is insufficient.
- any of the worker would run over `max_paid_instances_per_worker`.

If the deployment fails and the tokens are refunded (This is ensured by ink's revert on error). In cases where the deployment is successful, with a transferred value exceeding the necessary amount, any excess tokens will be refunded immediately.

**Note: The driver `sidevm_deployer` never check the existence of workers in the list. All workers listed will be considered when calculating the costs.**

Also, there are several configurations to sidevm instance are moved up to the driver to decide:
- memory size
- code size limit
- vital capacity (max gas between two host calls)

### CODE REVIEW

File: crates/pink/pink-extension/src/lib.rs [hunk 1]
```diff
@@ -161,6 +163,61 @@ pub enum PinkEvent {
...
+        deadline: u32,
+    },
+}
+
+#[derive(Encode, Decode, Serialize, Deserialize, Debug, Clone)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct SidevmConfig {
+    pub max_code_size: u32,
+    pub max_memory_pages: u32,
+    pub vital_capacity: u64,
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/pink/pink-extension/src/lib.rs:199]: Some comments added:
```rust
pub struct SidevmConfig {
    /// The maximum size of the code that can be uploaded to the SideVM instance.
    pub max_code_size: u32,
    /// The maximum number of memory pages(64KB per page) that can be allocated by the SideVM instance.
    pub max_memory_pages: u32,
    /// The max gas between two host function calls for the SideVM instance.
    pub vital_capacity: u64,
    /// The block number that the SideVM instance is allowed to run until.
    pub deadline: u32,
}
```

File: crates/phactory/src/system/mod.rs [hunk 2]
```diff
@@ -1834,6 +1860,53 @@ pub(crate) fn apply_pink_events(
...
+                        {
+                            Some(code) => SidevmCode::Code(code),
+                            None => SidevmCode::Hash(code_hash),
+                        };
+                        if let Err(err) = contract.start_sidevm(spawner, code, false, config)
+                        {
+                            error!(target: "sidevm", %vmid, ?err, "Start sidevm failed");
+                        }
+                    }
+                    SidevmOperation::SetDeadLine {
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/phactory/src/system/mod.rs]: Done

File: crates/pink/pink-extension/src/lib.rs [hunk 3]
```diff
@@ -161,6 +163,61 @@ pub enum PinkEvent {
...
+    pub deadline: u32,
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/pink/pink-extension/src/lib.rs:204]: Actually, it was rename from `live_until_block`. I feel some function names would be very verbose if using `live_until_block`.

File: crates/pink-drivers/sidevm_deployer/lib.rs [hunk 4]
```diff
@@ -7,27 +7,97 @@ use pink_extension as pink;
...
+        /// Owner of the contract.
         owner: AccountId,
         /// Contracts that are allowed to deploy sidevm.
         whitelist: Mapping<AccountId, ()>,
+        /// Price of sidevm instance per block per worker.
+        vm_price: Balance,
+        /// Price of memory per byte per block per worker.
+        mem_price: Balance,
+        /// Deadlines of paid instances indexed by workers.
+        paid_instances_by_workers: Mapping<WorkerId, Vec<BlockNumber>>,
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/pink-drivers/sidevm_deployer/lib.rs]: Renamed

File: crates/phactory/src/contracts/support.rs [hunk 5]
```diff
@@ -76,7 +78,10 @@ impl Decode for RawData {
 
 #[derive(Clone)]
 pub enum SidevmHandle {
-    Running(CommandSender),
+    Running {
+        cmd_sender: CommandSender,
+        stopped: WatchReceiver<bool>,
```
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/phactory/src/contracts/support.rs:83]: If the previous instance is stopping in progress, it is needed to wait until the state changed to stopped before we start the new instance.

File: crates/pink/pink-extension/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (h4x3rotab) [reply crates/pink/pink-extension/src/lib.rs:204]: Got it. Let's add some comments.

File: crates/pink/pink-extension/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (h4x3rotab) [reply crates/pink/pink-extension/src/lib.rs:202]: Trying to find a more straightforward name for "vital capacity":

- time_slice_budget
- max_gas_per_schedule
- schedule_gas_budget
- tick_gas_limit

[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/pink/pink-extension/src/lib.rs:202]: I think 'vital capacity' is the most accurate and expressive term for this concept. Here are my reasons:

- 'Vital capacity' is an illustrative term, which helps in fully understanding and remembering the scheduling principle of SideVM at a glance.
Why do I say this? Unlike conventional OS scheduling principles, in SideVM, when the gas runs out, the instance isn't scheduled away but is instead 'killed'. This is based on the underlying principle that it must run for a while before returning to the runtime to take a breath (literally taking a breath(吸一口gas)).
- It's more concise than the other descriptions.
- We already have a similarly concise error message, 'Stifled'.

Consider we are discussing something about this in a Chinese context:
"SideVM有个肺活量的设计..." (This easily brings to mind the complete picture of the design, without needing further explanation.)
"SideVM有个基于时间片调度的设计..."  (The audience might not easily recall the complete picture of the design, requiring further explanation.)

File: crates/pink/pink-extension/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/pink/pink-extension/src/lib.rs:204]: Already added.

**Review by h4x3rotab — APPROVED** [review]

**Pull request merged**